extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -796,47 +794,50 @@ \subsection{The definition of EUTXO-2}
 \begin{ruledfigure}{H}
   \center{
   \begin{tabular}{rll}
-  \s{CurrencyID}&: & an identifier for a custom currency. This is an alias
-  for the \s{Address} type\\
-                & & from Figure~\ref{fig:basic-types}.\\
+    \currency &: & an identifier for a custom currency. This is an alias
+                for the \s{Address} type\\
+             &  & from Figure~\ref{fig:basic-types}.\\
+    \\
+    \nativeCur &: & a distinguished \currency{} representing the native currency of the ledger.\\
+    \\
+    \token   &: & a type consisting of identifiers for individual tokens.  These will probably\\
+             &  & be hashes in practice.\\
     \\
-    \s{Token}&:& a type consisting of identifiers for individual tokens.  These will probably\\
-               & &be hashes in practice.\\
-\\
-    \qtymap &=& \Map{\s{CurrencyId}}{\Map{\s{Token}}{\s{Quantity}}}\\
+    \nativeTok &: & a distinguished \token{} representing the currency token of the native currency.\\
     \\
-  $\qtymap^{\pm}$ &=& \Map{\s{CurrencyId}}{\Map{\s{Token}}{\s{Quantity}^{\pm}}}\\
+    \qtymap   &=& \FinSup{\currency}{\FinSup{\token}{\qty}}\\
+    \\
+    \qtymappm &=& \FinSup{\currency}{\FinSup{\token}{\qtypm}}\\
   \end{tabular}
   }
   \caption{Extra basic types for the EUTXO-2 model}
   \label{fig:more-basic-types}
 \end{ruledfigure}
 
-\noindent The $\qtymap$ type represents a collection of funds from a
+\noindent The \qtymap{} type represents a collection of funds from a
 number of currencies and their subcurrencies, with all quantities
-being non-negative.  The $\qtymap^{\pm}$ type is similar, but
-negative quantities are allowed.
+being non-negative.  The \qtymappm{} type is similar, but
+negative quantities are allowed.  As with \qty{} and \qtypm{}, we regard
+\qtymap{} as a subtype of \qtymappm{} and convert freely between them.
 
 \begin{ruledfigure}{H}
   \[
   \begin{arraydefs}{rll}
     \s{Output}_2 &= (&\addr: \s{Address},\\
-    && \val: \qty\\
-    && \customvals: \qtymap,\\
+    && \val: \qtymap\\
```

Comment:
> at some point I asked whether we were allowed to have outputs containiat some point I asked whether we were allowed to have outputs containing zero quantities of currencies, and the conclusion was something like "Yes, why not? It might come in useful some time." This seems pretty reasonable: I can imagine that you might want to use a zero quantity to record the fact that you once owned a particular NFT, but not any more, or something along those lines.

> However, this means that we really do need to be able to record zero quantities in outputs: an output which contains {5 Groats, 0 Guilders} really is different from one that just contains {5 Groats}.

I definitely don't want this!

I think we misunderstood each other. When I said we should allow zero quantities I mean values that are zero as opposed to positive, not a variety of zeros distinguished by which currencies they have zero of. Just one zero!

Under the FSF interpretation all of these values:
- 5X
- 5X, 0Y
- 5X, 0Y, 0Z
- ...

are the same, and they're all equal to the value that has zero of *any* currency apart from X.

Reply:
Ah, that's the nub of this issue then (which is what I was trying to say above).  If we do allow 5X, 0Y to be different to 5X then I think we're forced to use finite maps (or something equivalent) to represent this in outputs, because FSFs can't distinguish the two.  We did discuss this before, and the consensus was that we should allow outputs with value zero, even zero Ada.  See

https://input-output-rnd.slack.com/archives/C21UF2WVC/p1557927092191100

(Ignore my stupid remarks about destroying things of value 0, and also check out what this Peyton Jones guy said).

I see that this was also discussed back in April 2018: https://input-output-rnd.slack.com/archives/C9DN2QE13/p1524107827000219

Maybe I'll raise it again, just to annoy everyone.

Reply:
I think we misunderstood each other in that conversation. What I was trying to say is:
- Yes, we should allow outputs where the value is zero.
- We need to think about how to handle different maps which all represent zero in our implementation. I was *not* saying that these *should* be distinguished at the spec level - the problem was about how to treat them the *same* in the implementation.

I think perhaps there was another confusion about the case where we destroy all of a currency X in some transaction. Do we need `X:0` in the output to know to run the MPS? No - that is keyed off the *forge* field, and if we are destroying a non-zero amount of X then it will appear in there. No problem.

Sorry if I seem to have sent you down a bit of a goose chase here...

Reply:
Yes.

Reply:
I'm still not grasping this.  Surely the spec has to capture the case when an output contains {5X, 0Y}, and it _must_ distinguish this from {5X}?  If we want to prove things about contracts on the basis of the specification then it has to model this case.  I don't see how FSFs do this.

Reply:
No, it *must not* distinguish those cases. They're the same. There is no situation where we want to distinguish them.

Forget everything that's come before: a value is just a thing that tells you how much there is for any currency/token. If we implement such a function with the maps `{5X, 0Y}` and `{5X}` then they give the same answers in all cases so they are the same.

Reply:
OK.  So should a contract/transaction/whatever be able to examine an input to check whether currency X is completely absent, or present with value 0?  I thought that the outcome of the earlier discussions was that it should be able to do that, and everything I've said above was based on that. If not, then I'm in full agreement with you.

Reply:
No, definitely not. I think we got confused between that and "outputs should be allowed to be zero", which is a bit different.

Sorry for not spotting this earlier, it looks like I'd have saved you some fuss!

Reply:
Let me ask on Slack what people's opinion is.

Reply:
Anyway, this is good!  It's shown that something requries clarification, at least as far as I'm concerned.  All the business with \mu was to deal with this, so if we can get rid of it then so much the better.

Reply:
OK, I'm convinced.  Disregard everything I've said on this subject!